### PR TITLE
feat (redirect): Create contact to reduce latency

### PR DESCRIPTION
**Pull request by nathan-moore-97**

### CODE REVIEW

File: qualtrix/api.py [hunk 1]
```diff
@@ -2,7 +2,11 @@
 qualtrix rest api
 """
 
+from asyncio import create_task
+import datetime
```
Comment (nathan-moore-97) [inline qualtrix/api.py]: Maybe redundant?

File: qualtrix/api.py [hunk 2]
```diff
@@ -29,7 +33,9 @@ class SessionModel(SurveyModel):
 
 class RedirectModel(SurveyModel):
     targetSurveyId: str
-    responseId: str
```
Comment (nathan-moore-97) [inline qualtrix/api.py:32]: add RulesConsentId back in, we need this for the contact list

File: qualtrix/api.py [hunk 3]
```diff
@@ -46,15 +52,53 @@ async def get_response(request: ResponseModel):
 
 
 @router.post("/redirect")
-async def get_redirect(request: RedirectModel):
+async def intake_redirect(request: RedirectModel):
```
Comment (nathan-moore-97) [inline qualtrix/api.py:60]: I really doubt we will ever go back to the 1.0 method for any reason. I feel safe moving this entirely, and there is no real "retry" scheme i can think of that would benefit the current implementation. It is a single operation with a number of side effects. I am planning to refactor this to be more atomic and put those susbsequent steps not directly related to the workflow on the client side.

File: qualtrix/api.py [hunk 4]
```diff
@@ -46,15 +52,53 @@ async def get_response(request: ResponseModel):
...
+            settings.LIBRARY_ID,
+            settings.INVITE_MESSAGE_ID,
+            settings.MAILING_LIST_ID,
+            request.targetSurveyId,
+        )
+        link = client.get_link(request.targetSurveyId, email_distribution["id"])
+
+        # If link creation succeeds, create reminders while the link is returned
+        create_task(create_reminder_distributions(email_distribution["id"]))
+        create_task(add_user_to_contact_list(link["link"], directory_entry["id"]))
```
Comment (nathan-moore-97) [inline qualtrix/api.py]: We can fire and forget both these tasks. They are side effects of the Rules Survey completion and are unrelated to the redirect flow (barring consuming some of its return values)

Putting these two calls in another endpoint will be the primary target for redirect 2.1

File: qualtrix/main.py [hunk 5]
```diff
@@ -12,6 +12,18 @@
 
 app = fastapi.FastAPI()
 
+origins = ["*"]
+
+from fastapi.middleware.cors import CORSMiddleware
```
Comment (nathan-moore-97) [inline qualtrix/main.py]: Remove

File: qualtrix/settings.py [hunk 6]
```diff
@@ -32,11 +49,33 @@
...
+        REMINDER_MESSAGE_ID = config["reminder_message_id"]
+        INVITE_MESSAGE_ID = config["invite_message_id"]
+        MAILING_LIST_ID = config["mailing_list_id"]
+        FROM_EMAIL = config["from_email"]
+        REPLY_TO_EMAIL = config["reply_to_email"]
+        FROM_NAME = config["from_name"]
+        INVITE_SUBJECT = config["invite_subject"]
+        REMINDER_SUBJECT = config["reminder_subject"]
+        SURVEY_LINK_TYPE = config["survey_link_type"]
+        DEMOGRAPHICS_SURVEY_LABEL = config["demographics_survey_label"]
```
Comment (nathan-moore-97) [inline qualtrix/settings.py:65]: Do we want to create a more structured config? all of these variables at global scope in the config seems fine to me

Comment (dzaslavskiy) [reply qualtrix/settings.py:65]: ok

**Review by dzaslavskiy — APPROVED** [review]

**Review by dzaslavskiy — APPROVED** [review]

**Pull request merged**